### PR TITLE
Restore stdio in test rig

### PR DESCRIPTION
**Pull request by som-snytt**
Save and restore stdio in the test rig child process.

Also add a test from the duplicate ticket for 23245.

### CODE REVIEW

**Comment by hamzaremmal** [comment]
> Maybe there's no more skipping...

There is no more skipping.

**Comment by som-snytt** [comment]
Added a bit of output for more stack trace. The test will always fail, to make re-running easier.

My first guess was NPE in clinit. My second guess was a bad "nullable field" getting nulled after lazy val init; perhaps the outer pointer itself, which was used only for the lazy init. But the outer is not a class param in the example; its rhs is just a static module. Maybe the trait init? but that is empty. Maybe a class isn't initialized? Maybe a missing inner class, except the module is top level.

I could not replicate locally (WSL) but interacting with the test rig made me want to contribute improvements, as I intended.

I see that the flaky test is not included in the group with `-Ylegacy-lazy-vals`, which was my fourth guess.

**Comment by som-snytt** [comment]
Woo-hoo!
```
Output from 'tests/run/i23245a' did not match check file. Actual output:
java.lang.NullPointerException: Cannot invoke "java.io.PrintStream.println(String)" because "java.lang.System.err" is null
	at logadapter.LogAdapter.info(api.scala:23)
	at Test$.main(test_2.scala:3)
	at Test.main(test_2.scala)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at dotty.tools.vulpix.ChildJVMMain.runMain(ChildJVMMain.java:40)
	at dotty.tools.vulpix.ChildJVMMain.main(ChildJVMMain.java:47)
```
Oh wait, does that make any sense? I'll look at it later.

**Comment by som-snytt** [comment]
The test for https://github.com/scala/scala3/pull/2781 was aware that nulling stdio streams is unfriendly, but did it anyway, in the interests of science.

Any test thereafter in the same runner JVM that used `System.err`, including the default exception handler in doing `e.printStackTrace()`, would throw NPE.

This doesn't seem to be a rampant problem, but for safety, the `ChildJVMMain` that communicates with the test rig now saves and restores stdio.

Equivalent in `partest`:

https://github.com/scala/scala/blob/v2.13.18/src/partest/scala/tools/partest/nest/Runner.scala#L242

File: compiler/test/dotty/tools/vulpix/ChildJVMMain.java [hunk 1]
```diff
@@ -41,11 +41,27 @@ private static void runMain(String dir) throws Exception {
...
-      }
+        while (true) {
+            var err = System.err;
+            var out = System.out;
+            var in  = System.in;
+            try {
+              runMain(stdin.readLine());
+            }
+            finally {
+              if (err != System.err) {
```
Comment (hamzaremmal) [inline compiler/test/dotty/tools/vulpix/ChildJVMMain.java]: why branch here? we can just call `System.setErr` regardless of if it changed or not.

Comment (som-snytt) [reply compiler/test/dotty/tools/vulpix/ChildJVMMain.java]: Just to be finicky. I needed a block for a println while verifying, and then I decided to avoid any side-effects, however benign, of calling the setter.

Most of my time was spent formatting the Java code.

Worth adding that this class could be written in Scala.

Comment (som-snytt) [reply compiler/test/dotty/tools/vulpix/ChildJVMMain.java]: Looks like I have an opportunity edit it again anyway. I requested your review because of your proximate interest in the build passing correctly. By proximate, I mean near and dear to your heart.

**Comment by som-snytt** [comment]
Idea for `MainProxies` is to use default args as fallback for missing args. Inspired by test that needs args.

**Comment by Gedochao** [comment]
@som-snytt You know the drill - still working on this? `stat:tbc`?

**Comment by som-snytt** [comment]
This was for the NPE in https://github.com/scala/scala3/issues/24572

Edit: that is, it should reproduce the test failure at some point. (Unless it doesn't.)

But there was another NPE I couldn't get to the bottom of one day:
That was `SignatureBuilder` https://github.com/scala/scala3/issues/24183

**Comment by som-snytt** [comment]
The one failing test passes locally, either as test or task.
```
sbt:scala3-bootstrapped> scalac -indent -Yno-double-bindings -Yforce-sbt-phases -Xsemanticdb -Xverify-signatures -pagewi
dth  120 -color:never -Xtarget 17 -Ycheck:all -Wconf:id=E222:s -Ycheck:instrumentCoverage  -coverage-out /tmp/coverage00
 tests/pos-deep-subtype/i21015.scala
[info] running (fork) dotty.tools.dotc.Main -d /home/amarki/projects/scala3/out/default-last-scalac-out.jar -classpath /home/amarki/projects/scala3/library/target/scala-library-bootstrapped/scala-library-3.8.5-RC1-bin-SNAPSHOT.jar -indent -Yno-double-bindings -Yforce-sbt-phases -Xsemanticdb -Xverify-signatures -pagewidth 120 -color:never -Xtarget 17 -Ycheck:all -Wconf:id=E222:s -Ycheck:instrumentCoverage -coverage-out /tmp/coverage00 tests/pos-deep-subtype/i21015.scala
```

Edit: it fails on another PR as well. It's an "independent" failure.

**Comment by som-snytt** [comment]
Locally
```
diff --git a/tests/neg/i22890.check b/tests/neg/i22890.check.out
index f5e37b10da..b92ad4359c 100644
--- a/tests/neg/i22890.check
+++ b/tests/neg/i22890.check.out
@@ -1,3 +1,6 @@
+package scala contains object and package with same name: caps.
+This indicates that there are several versions of the Scala standard library on the classpath.
+The build should be reconfigured so that only one version of the standard library is on the classpath.
 -- [E161] Naming Error: tests/neg/i22890/caps_1.java:3:0 ---------------------------------------------------------------
 3 |class caps { // error: caps is already defined as package caps
   |^
```
Plus usual cruft on JVM 25, I thought a PR took care of that.

**Comment by Gedochao** [comment]
@som-snytt this seems to be in need of a rebase, before I find a reviewer :thinking:

**Comment by som-snytt** [comment]
It's hard to keep up.

This would be an opportunity to add the "command audit" so that when a test fails, you could see the history of what the process has been doing.

My other comment was that it needs a dose of "less fewer braces".

This was precursor to more test debugging and so-called "hardening".

**Comment by som-snytt** [comment]
That's what I meant by my previous comment: I had trouble motivating the changes.

I'll split it into pieces that at least correspond to a real need, and also roll back "style".

Nothing here will rock anyone's world. The failing test didn't fail again. "Yay."

Thanks for looking.

**Comment by SolalPirelli** [comment]
@som-snytt I'm fine merging this as-is, fwiw, it seems to move the code in the right direction overall even if that direction is not super clear

**Comment by som-snytt** [comment]
Squashed some commits and rebased.

I had already done some style tweaking, such as indented case blocks.

I did not supply braces that were previously missing. One style note is that a top-level element that spans the file does not need braces: indeed, colon announces that structure. 

The useful changes are stdio security, summarize skipped tests, mutation in RunnerOrchestration, ChildJVMMain doesn't need to be in Java (I didn't check the history).

I take it from my previous comment that I intended for `MainProxies` to use default args when command line args are not supplied, so I prefer to save the refactor for when I do that (or if). I dropped that commit. (Now all changes are only under test.)

It's really terrible that the i23245a error did not recur, but there are other tests that incur flakiness, and which may just be a side effect of how the tests are run under the test rig. Is it that a JVM "runner" hits an error condition (OOM, SO) that affects a subsequent test? A memory leak? etc. Hard to diagnose without reproducibility.

**Comment by mbovel** [comment]
That PR significantly slowed down `test-scala3-compiler-bootstrapped`. Is that expected?

Date       | Commit     | Duration
-----------|------------|---------
2026-04-02 | 6c411a6ea6 | 53.3 min
2026-04-02 | 5ad705f658 | 53.1 min
2026-04-02 | a16b497ced | 53.7 min
2026-04-02 | c39a78e9df | 55.2 min
2026-04-02 | ecf8e5c772 | 54.0 min
2026-04-03 | 68d439f5c9 | 54.0 min
2026-04-03 | a7f6cae5f6 | 53.6 min
2026-04-04 | 2a79b8e121 | 55.9 min
2026-04-04 | 9478256a04 | 54.3 min
2026-04-05 | 2c668b5a0b | 54.9 min
2026-04-06 | ec88c95f67 | 55.2 min
2026-04-07 | cc7d6db700 | 53.8 min
2026-04-07 | 51397c224d | 53.1 min
2026-04-07 | e3f9fc1a6c | 55.8 min
2026-04-07 | cc4a3e6fac | 52.2 min
2026-04-08 | 4da95f175b | 51.8 min
2026-04-08 | f1005e2a4b | 51.9 min
2026-04-08 | 5feaf5ae80 | 54.0 min
2026-04-08 | df7d8762a7 | 53.3 min
2026-04-08 | c23f2348f6 | 54.1 min
2026-04-09 | 2885468769 | 53.6 min
2026-04-09 | 9c90195684 | 54.7 min
2026-04-09 | 91e097e6c2 | 54.7 min
2026-04-09 | e679e6ad32 | 70.1 min
2026-04-10 | 77308197e9 | 66.8 min
2026-04-10 | 1610a73976 | 69.9 min
2026-04-10 | 20fc4c7576 | 67.3 min
2026-04-10 | c1a14eed77 | 70.3 min

**Comment by mbovel** [comment]
> That PR significantly slowed down `test-scala3-compiler-bootstrapped`. Is that expected?

@claude says we're missing:

```diff
diff --git a/compiler/test/dotty/tools/vulpix/RunnerOrchestration.scala b/compiler/test/dotty/tools/vulpix/RunnerOrchestration.scala
index a5bac56b19..e3fc684edd 100644
--- a/compiler/test/dotty/tools/vulpix/RunnerOrchestration.scala
+++ b/compiler/test/dotty/tools/vulpix/RunnerOrchestration.scala
@@ -215,7 +215,9 @@ trait RunnerOrchestration:
     }
 
     private def discardRunner(runner: Runner): Unit = synchronized {
+      runner.kill()
       busyRunners -= runner
+      notify()
     }
 
     private def withRunner(op: Runner => Status)(using SummaryReporting): Status =
```

which keeps zoombie runners taking CPUs. Would that make sense?

**Comment by mbovel** [comment]
> Would that make sense?

I'll just try running the CI with the proposed fix, that should be easy to validate or invalidate: https://github.com/scala/scala3/pull/25764.

**Comment by som-snytt** [comment]
I can tell from the fan by running the vulpix tests. I'll confirm the claude and review that commit.
